### PR TITLE
TRANSFORM: the pipelined sweep, object-storage roots, and their review

### PR DESCRIPTION
**Pull request by vboussot**
## Summary

TRANSFORM's streamed sweep now decomposes a sheared landing in blocks instead of full-plane slabs, folds a chain's pull maps before the first read, and overlaps the read, the chain and the write: one block ahead on a producer thread, one behind on a single writer, a deeper queue only while the tile stays the tile the output store is chunked by. The rank's cores are divided once (`rank_cpu_share()`) between torch, ITK and zarr, a region is assembled on the rank's pool, and the decoded-chunk cache is bounded by the declared per-rank budget, min(budget, max(256 MiB, budget/3)), in TRANSFORM, PREDICTION and EVALUATION alike, set on the rank where a spawned process starts. A sweep declares its source windows (`Dataset.plan_region_reads`), so the cache evicts the chunk whose next declared use is furthest away and ranks an undeclared chunk by its recency. A `dataset_filenames` entry may name a URI: `:omezarr` serves a remote root read-only, credentials come from fsspec's own environment (`FSSPEC_S3_ANON=true konfai TRANSFORM --config Transform.yml`), `pip install konfai[s3]` is the new extra, every probe answers for both kinds of root or raises (`konfai/utils/uri.py`), a remote Write is the REFUSED verdict at the plan, and the cohort walk asks the root only for the cases a subset names. The host resample hands SimpleITK its grid and copies the region once, `Clip` and `FlatLabel` fill through the mask, `Gradient` writes into one buffer, a device block lands in a reused host slot, and the sweep prints its own clock line with the residual. A chain on a dtype torch has no kernel for (uint16) fails naming `TensorCast` as the remedy. The review commits fix: the resident blocks a pipelined sweep is priced at (`_sweep_resident_slabs`, 2 + d + 2), stale chunks after a whole-volume rewrite, a sweep kept on one thread when its Save lands on the `:h5` file it reads, a forked DataLoader child rebuilding the rank pool, a read plan retired at its last read, zarr's concurrency kept whole on a share of a few cores, `torch.median` widening kept where torch has no kernel, and a sweep's publish charged to the write phase.

## Measured

ExaSPIM rows: 513x1331x1776 uint16 in 256^3 chunks resampled through a stored rigid+affine, `konfai TRANSFORM --config Resample.yml -y [--gpu 0]`, memory_budget 4G, before -> after of that step.

| Step | Measurement |
|---|---|
| Sweep in blocks, not full-plane slabs | source voxels requested 2168.8 M -> 1292.0 M (1.79x -> 1.07x); zarr chunks decoded 168 -> 144 (3.63 GB -> 2.83 GB); gpu fast 10.4 s 6.90 GB -> 8.1 s 5.18 GB; gpu exact 11.7 s 6.86 GB -> 10.1 s 5.25 GB; cpu exact 17.3 s 10.40 GB -> 12.8 s 7.33 GB; cpu 1 thread 36.8 s 10.47 GB -> 24.9 s 7.31 GB |
| `rank_cpu_share()` divided once, cache bounded by the declared budget | gpu fast 8.1 s 208% 5.18 GB -> 8.3 s 243% 4.33 GB; gpu exact 10.1 s 188% 5.25 GB -> 10.2 s 219% 4.37 GB; cpu exact 12.8 s 411% 7.33 GB -> 13.3 s 435% 6.66 GB; a run declaring 4G had peaked at 6.90 GB; zarr at the whole share 6.7 s, half 6.4 s, a third 6.1 s, a sixth 6.2 s |
| Read one block ahead, write one behind | gpu fast 8.3 s 243% 4.33 GB -> 6.8 s 286% 3.56 GB; gpu exact 10.2 s 219% 4.37 GB -> 8.6 s 258% 3.48 GB; cpu exact 13.3 s 435% 6.66 GB -> 13.8 s 430% 4.33 GB; the GPU path is worth 18 %, the host path pays 4 % |
| Chain off the reading thread (alternating A/B) | gpu fast 9.6 s 253% -> 6.9 s 336%; cpu exact 14.7 / 14.9 / 16.3 s ~427% -> 10.8 / 11.1 / 11.5 s ~614%; OMP_NUM_THREADS=1 35.5 / 37.8 / 37.4 s -> 35.9 / 37.5 / 36.1 s |
| `Dataset.plan_region_reads`, evict by next declared use | chunks decoded 186 -> 148, bytes decoded 4.32 GB -> 3.16 GB; gpu fast 6.3 / 6.5 / 6.5 s -> 5.9 / 5.7 / 5.6 s; cpu exact 11.7 / 11.9 / 12.6 s -> 12.4 / 12.5 s; simulated on 40 blocks over 126 chunks with a cache of 39: 173 decodes under LRU, 133 optimal, 126 the floor |
| Resampler handed its grid, one cast and one copy (host path, quiet machine, against f0d497d) | 9.8 / 10.5 / 10.0 s 600% CPU 4.33-4.52 GB -> 7.8 / 7.8 s 705% CPU 3.90-3.97 GB -> 6.2 / 6.2 s 636% CPU 3.56 GB; the chain 9.1 -> 4.9 s; standalone, 40 blocks of 384^3, 5.52 s -> 1.90 s |
| Device block landed in a reused slot, then a deeper queue | device 5.5 / 5.7 s -> 5.1 / 5.2 s, fetch phase 1.7 s -> 0.9 s; then device 5.1 / 5.2 s -> 5.0 / 5.0 s, host 6.5 / 6.5 s -> 6.2 / 6.2 s, tile unchanged at 384^3 in 40 blocks |
| Remote root, two brains read anonymously from `s3://aind-open-data`, output local, `konfai TRANSFORM --config Resample.yml -y --gpu 0` | plan (headers only) 23 s; run 19:12 wall, 1130.9 s in the materialisation loop; downloaded 1.898 GB; peak RSS 5.15 GB; over a 1.6 MB/s link, 2.573 GB in 38:34 with the cache bounded to a third of a declared 4 GiB |
| Cohort walk asks the root for the named cases (64 cases per root, `tests/unit/test_cohort_walk.py`) | entry opens for one named case: mha 64 -> 2, omezarr 64 -> 2; no subset and a slice stay at 64; h5 stays at 1 |
| Peak memory around a stage (VmHWM; CUDA peak for the grid_sample slab) | Clip +79 MB -> +0 MB and FlatLabel +1603 MB -> +0 MB on a 384^3 int16 block; Gradient on a 256^3 float32 block, flattened +593 MB -> +266 MB (working_multiple 7.8 -> 3.0), per_dim +395 MB -> +200 MB (2.9 -> 0.0); a 3-channel 128x512x512 float32 transform slab +1572 MB -> +786 MB; InferenceStack median, 5-case float16 stack, 7.99 GB -> 3.07 GB; Mean first accumulate, int16 case, 1.23 GB -> 0.62 GB; 4-channel 192x640x640 uint16 host resample 6.46 GB -> 5.23 GB; grid_sample CUDA peak on a 64x384x384 slab 0.642 GB -> 0.340 GB |
| Sweep clock line, always on | phase() enter+exit 555 ns, 7 phases x 40 blocks = 156 us per sweep |
| Review: blocks a pipelined sweep holds (`.scratch/inflight_peak.py`, three runs each) | depth 0: 2 slabs, priced 2 (was 2); depth 1: 5, priced 5 (was 4); depth 2: 6, priced 6 (was 6); depth 3: 7, priced 7 (was 8) |
| Review: cache capacity on the plan | 96 KiB budget: 96 KiB (was 256 MiB, 2731x the budget); 512 MiB: 256 MiB; 3 GiB: 1 GiB |
| Review: undeclared chunks by recency (`.scratch/belady_variants.py 4 8 20 39 60`; decodes LRU / never-again / by recency) | cache 39: 560 / 604 / 557; cache 20: 976 / 748 / 725; cache 60: 532 / 465 / 437; on the ExaSPIM bench, 6af1202 against this branch, 148 chunks decoded in 44 array reads -> 138 in 32 |
| Zarr concurrency re-measured under the split-thread sweep, 24 cores (host / device) | conc 4 12.1 / 13.4 s, 5.1 / 5.2 s; conc 8 10.7 / 11.1 s, 4.9 / 5.1 s; conc 12 11.0 / 11.3 s, 4.7 / 5.1 s; conc 24 11.0 / 12.3 s, 5.3 / 5.4 s; a third of the share stands |

**Values.** Against the undecomposed pass over three z slices, max |diff| on a 0-35465 range: exact 2 -> 1, fast 6 -> 2 after the block sweep; the pipelined loop is bit-identical on the host path and max |diff| 1 of 35465 on the device path; the helpers refactor max |diff| 3 of 35465 on the fast path, the same 23398 voxels. The resampler's grid and the one copy are checked against f0d497d on three z slices over 0..44889, the reused slot, the deeper queue and the per-axis grid on both paths: max|diff| 0, 0 of 7091568 voxels differing. Clip, FlatLabel, Gradient, the median pick and the slab cast are bit-identical over the dtypes their bodies list. The pricing, cache-bound, refusal and publish-timing commits move no values: a hit serves the bytes a decode would, and the tile is the same tile.

**Tests.** New: `tests/unit/test_sweep_pipeline.py`, `test_sweep_tiling.py`, `test_cohort_walk.py`, `test_remote_dataset.py`. Extended: `test_memory_budget.py`, `test_runtime.py`, `test_data_manager.py`, `test_transform.py`, `test_transformer_workflow.py`, `test_ome_zarr_data_surface.py`. The steps whose bodies name a per-module `.pixi/envs/dev/bin/python -m pytest` ran it, then `pixi run test-fast`; `pixi.lock` is relocked for the s3 extra.

**Stack.** Based on `main` (e73a6af), the first PR of the 1.8.2 stack; `pr/1.8.2-02-storage` is stacked on it. Before deleting this branch, retarget that PR.

## Summary by CodeRabbit

* **New Features**
  * Added read-only support for OME-Zarr datasets hosted at S3-compatible URI locations.
  * Added optional S3 installation support and credential configuration guidance.
  * Improved streaming with memory-aware tiling, bounded read-ahead, write-behind, and timing details.
  * Added efficient selection of explicitly named dataset cases.
* **Bug Fixes**
  * Improved decoded-chunk cache limits and memory-budget enforcement.
  * Added clearer errors for unreachable datasets, unsupported data types, and refused remote writes.
  * Improved data-type preservation during resampling and inference reductions.
* **Documentation**
  * Clarified memory usage, RAS orientation, remote storage, streaming behavior, and planning output.

### CODE REVIEW

File: docs/source/concepts/streaming.md [hunk 1]
```diff
@@ -80,6 +80,13 @@ The figure is an estimate, computed from file headers (`prod(shape) × 4 bytes`
 per group), so it ignores the dtype you stored, size-changing transforms and
 augmented copies. It is a switch, not an allocator limit.
 
+It also bounds only the buffers the pipeline holds for your data, the OME-Zarr
+decoded-chunk cache among them (a third of the budget, in every workflow). The
+interpreter, torch and its CUDA context, the model and each worker's own
+working set sit outside it, so the peak RSS a run reports is the budget plus a
+floor that does not move when the budget does: lowering the budget lowers the
+peak by roughly what you took off, never to the budget itself.
+
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/concepts/streaming.md:89]: Corrected further up the stack: #137 rewrites this sentence to state the share, the floor and the cap together ("a third of the budget, never more than the budget, and never below a 256 MiB floor unless the budget itself is smaller"), because that PR is where the cache's share of a declared budget is bounded. Nothing left to change here.

File: konfai/transformer.py [hunk 2]
```diff
@@ -626,10 +667,13 @@ def _plan_reduction(self, item: WorkItem, overwrite: bool) -> TransformPlanEntry
...
         reduction_plan = reduction.plan()
-        skipped = not item.pending(reduction.reduce.output, overwrite)
+        remote = self._remote_destination(item.manager, reduction.destination)
+        skipped = remote is None and not item.pending(reduction.reduce.output, overwrite)
         verdict = Verdict.SKIP if skipped else (Verdict.REDUCE if reduction_plan.streams else Verdict.REFUSED)
         reason = reduction_plan.refusal if not reduction_plan.streams else "\n".join(reduction_plan.body_lines())
-        if verdict is Verdict.REDUCE:
+        if remote is not None:
+            verdict, reason = Verdict.REFUSED, remote
+        elif verdict is Verdict.REDUCE:
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/transformer.py:676]: Fixed. `_remote_destination` now always scans the chain's own Saves, and what a route names is checked beside them, so a chain such as `Clip -> Save: memory://bucket/cache -> Reduce -> Write: ./out` is REFUSED at the plan instead of reading every member first. Pinned by test_a_reduction_chain_is_refused_for_a_remote_save_before_its_terminal_write, which reports REDUCE without the change.

File: konfai/utils/dataset.py [hunk 3]
```diff
@@ -2590,52 +2631,86 @@ def __init__(
...
-        case = next((child for child in base.iterdir() if child.is_dir()), None)
-        if case is None:
+        if not uri.is_dir(root):
             return None
-        for entry in sorted(case.iterdir()):
+        if uri.is_uri(root):
+            # Only the store backend reads a remote root, and a store is told by its name: a
+            # remote entry is never probed as a path, which on a bare name asks the working directory.
+            names = Dataset._first_case_entries(root)
+            return "omezarr" if any(is_store_name(name.name) for name in names) else None
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/dataset.py:2670]: Fixed. The accepted set is published once (`STORE_FORMS`, longest first so a compound extension wins over its tail) and both the resolution and the listing take it, so all five spellings resolve. Pinned by test_a_store_named_in_any_accepted_spelling_resolves_and_lists, which fails on two of the five without the change.

File: konfai/utils/uri.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            f"No filesystem is installed for '{protocol}://' ({exc}).",
+            f"Install it with: pip install konfai[{extra}]"
+            if extra
+            else f"Install the fsspec implementation for {protocol}.",
+        ) from exc
+    except (TypeError, ValueError) as exc:
+        raise DatasetManagerError(
+            f"'{protocol}://' refused its fsspec configuration: {exc}.",
+            "Check the FSSPEC_" + protocol.upper() + "_* variables and ~/.config/fsspec/.",
+        ) from exc
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/uri.py:107]: Fixed. An unregistered scheme and a configuration that failed now say which one happened: fsspec's "Protocol not known" no longer sends the reader to FSSPEC_<PROTO>_* variables that do not exist for a protocol it has never heard of. Pinned by test_an_unregistered_scheme_is_not_reported_as_a_configuration_to_check.

File: tests/unit/test_transform.py [hunk 5]
```diff
@@ -1003,3 +1003,14 @@ def test_a_streamed_mha_mask_is_read_by_region_and_never_held_whole(tmp_path: Pa
...
     np.testing.assert_array_equal(whole[0, :, 0, 0].numpy(), [-1, -1, 1, 1, 1, 1, -1, -1])
+
+
+@pytest.mark.parametrize("dtype", [torch.uint16, torch.bool, torch.uint8, torch.int16, torch.float16])
+def test_inference_stack_median_picks_the_same_element_whatever_the_dtype(dtype: torch.dtype) -> None:
+    """torch has no median kernel for uint16, uint32, uint64 or bool: the stack is widened for those
+    alone, and the element picked is the one the widened median picks."""
+    stack = torch.tensor([[7, 1], [3, 5], [9, 0]]).to(dtype)
+    expected = torch.median(stack.float(), dim=0).values.to(dtype)
+    reduced = InferenceStack(dataset="", name="pred", mode="median")._reduce(stack)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_transform.py:1014]: Both call sites now go through monkeypatch. The test that names them clears the whole workflow environment on purpose (monkeypatch.delenv over _WORKFLOW_ENV, so it can watch configure_workflow_environment repopulate it), which is why the mode was being set again by hand rather than left to the autouse fixture in tests/conftest.py. Setting it through monkeypatch keeps that intent and makes the write undoable.

test_transform.py needs nothing: the autouse _konfai_env fixture already sets KONFAI_config_file and KONFAI_CONFIG_MODE for every test, and nothing in that file clears them.

**Comment by vboussot** [comment]
On the outside-diff note about patching KONFAI_config_file in the two helpers: monkeypatch restores the value it recorded on first contact, whatever is written afterwards, and configure_workflow_environment is production code whose write cannot be intercepted from a test.

    os.environ["KONFAI_config_file"] = "ORIGINAL"
    m.setenv("KONFAI_config_file", "fixture-value")   # the autouse fixture
    m.delenv("KONFAI_config_file")                    # what this test does
    os.environ["KONFAI_config_file"] = "written by production code"
    m.undo()
    # -> ORIGINAL

So no value can leak across tests here. The config mode was different: the test clears it deliberately and the workflow helper does not set it, so it had to be set again, and that is now done through monkeypatch.
